Yoast/wordpress-seo
# 19988 crawl cleanup permalink unit tests

## Context


* Add unit test for the crawl cleanup permalinks class.

## Summary


This PR can be summarized in the following changelog entry:

* Refactors `Crawl_Cleanup_Permalink` class and adds unit tests.

## Relevant technical choices:

* Added a `Crawl_Cleanup_Helper`.

## Test instructions

### Test instructions for the acceptance test before the PR gets merged
This PR can be acceptance tested by following these steps:

*

#### Relevant test scenarios


### Test instructions for QA when the code is in the RC




QA can test this PR by following these steps:

* Go to `Yoast SEO`->`Settings`->`Advanced`->'Crawl optimisation`
* Enable `Remove unregistered URL parameters` 
* Add parameter to `Additional URL parameters to allow`, for example `allowed_parameter`.
* Go to front page with an unknown parameter, for example: `https://www.example.com/?unknown_parameter=yes`.
* Check that you are **not** redirected and the unknown parameter is still there.
* Log out and go to the same url (`https://www.example.com/?unknown_parameter=yes`), check you are redirected.
* Add the allowed parameter instead and check you are **not** redirected  (`https://www.example.com/?allowed_parameter=yes`)
* Go to `Settings` -> `Reading` and switch the `Your homepage displays` to `A static page`
* Select one page to be the homepage, and one page to be the `Posts page`.
* When logout, check these pages with `?unknown_parameter=yes` are redirected:
  -  `Posts page`
  - Home page.
  - Single post/page/CTP
  - Search result page with search parameter for example `?s=test`
  - Page 404
  - Category/Tags/ Taxonomy page.
  - In a category page that has more than one posts (and in the site's Settings->Reading settings, put `Blog pages show at most` as `1`, go to the 2nd page), go to the 2nd page of that category. It will most probably be a URL ending with `/category/category-slug/page/2/`. Append to that `?unknown_parameter=yes`
* Follow the same steps with both allowed and unknown parameters for example: `https://www.example.com/?unknown_parameter=yes&allowed_parameter=test` and check you are redirected but only the unknown parameter is removed.

( Related https://github.com/Yoast/wordpress-seo-premium/pull/3573 )

## Impact check

This PR affects the following parts of the plugin, which may require extra testing:

*

## UI changes


## Other environments


## Documentation


## Quality assurance


## Innovation


Fixes [[Crawl cleanup: enhance test coverage for crawl-cleanup related integrations](https://github.com/Yoast/wordpress-seo/issues/19988)](https://github.com/Yoast/wordpress-seo/issues/19988)
